src/base/static/components/input-form/index.js

```diff
@@ -0,0 +1,551 @@
+              constants.MAP_DRAWING_TOOLBAR_TYPENAME &&
+            val.get(constants.FIELD_STATE_FIELD_TYPE_KEY) !==
+              constants.ATTACHMENT_FIELD_TYPENAME
+        )
+        .map(val => val.get(constants.FIELD_STATE_VALUE_KEY))
+        .toJS()
+    );
+
+    this.richTextFields.forEach(fieldName => {
+      // Replace image data with placeholders built from each image's name.
+      attrs[fieldName] = attrs[fieldName].replace(
+        /<img.*?name="(.*?)".*?>/g,
+        constants.RICH_TEXT_IMAGE_MARKUP_PREFIX +
+          "$1" +
+          constants.RICH_TEXT_IMAGE_MARKUP_SUFFIX
+      );
+    });
+
+    this.attachments.forEach(attachment => {
+      model.attachmentCollection.add(attachment);
+    });
+
+    // Save autofill values as necessary.
+    selectedCategoryConfig.fields
+      .concat(this.props.placeConfig.common_form_elements)
+      .forEach(fieldConfig => {
+        if (fieldConfig.autocomplete) {
+          Util.saveAutocompleteValue(
+            fieldConfig.name,
+            this.state.fields
+              .get(fieldConfig.name)
+              .get(constants.FIELD_STATE_VALUE_KEY),
+            constants.AUTOFILL_DURATION_DAYS
+          );
+        }
+      });
+
+    if (this.customGeometryFieldname) {
+      attrs.geometry = this.state.fields
+        .get(this.customGeometryFieldname)
+        .get(constants.FIELD_STATE_VALUE_KEY)
+        .get("geometry");
+      attrs.style = this.state.fields
+        .get(this.customGeometryFieldname)
+        .get(constants.FIELD_STATE_VALUE_KEY)
+        .get("geometryStyle");
+    } else {
+      const center = this.props.map.getCenter();
+      attrs.geometry = {
+        type: "Point",
+        coordinates: [center.lng, center.lat],
+      };
+    }
+
+    // Fire pre-save hook.
+    // The pre-save hook allows flavors to attach arbitrary data to the attrs
+    // object before submission to the database.
+    if (this.props.customHooks && this.props.customHooks.preSave) {
+      attrs = hooks[this.props.customHooks.preSave](attrs);
+    }
+
+    model.save(attrs, {
+      success: response => {
+        Util.log("USER", "new-place", "successfully-add-place");
+        this.setState({ isFormSubmitting: false });
+
+        // Fire post-save hook.
+        // The post-save hook allows flavors to hijack the default
+        // route-to-detail-view behavior.
+        if (this.props.customHooks && this.props.customHooks.postSave) {
+          this.props.customHooks.postSave(
+            response,
+            model,
+            this.defaultPostSave.bind(this)
+          );
+        } else {
+          this.defaultPostSave(model);
+        }
+      },
+      error: error => {
+        Util.log("USER", "new-place", "fail-to-add-place");
+      },
+      wait: true,
+    });
+  }
+
+  defaultPostSave(model) {
+    this.resetState();
+    this.props.router.navigate(Util.getUrl(model), { trigger: true });
+  }
+
+  getSpecialFieldProps(fieldType) {
```

Comment:
See my comment about having specific form field logic in our InputForm.

In this case, I think it's reasonable to just pass our `map`, `router`, `mapConfig`, etc props to the FormField, and dealing with the specific form field types there.

Whatever optimization this is bringing us, I just don't think it's worth the complexity and it is also blurring our InputForm abstraction.

Eventually, we should find ways to avoid having to pass these props around. For example, I don't think we need to do that for `emitter` (see my other comment in GeocodingField). There are probably some changes we can make to avoid doing this with the other props - happy to chat about that later.

Reply:
Ok, sounds good. I tried to err on the side of limiting props to components, but I agree this `getSpecialFieldProps` seems bloated.